File: python/distributed_runtime/activation_integrity.py
"""Commit-first activation sketches for heterogeneous stage verification.

Adapted and substantially modified from ``shard/challenge.py`` at revision
fcf728096948c7686bcf0897e9acb75d1abda1d5 (Apache-2.0). Mycellios uses a
domain-separated hash stream for projection indexes, strict bounded wire
validation and a Torch-free verifier. Tensor production imports Torch lazily.

This module supplies a verification primitive. It does not by itself prove a
physical multi-host run or authorize settlement.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import math
import secrets
import sys
from collections.abc import Mapping, Sequence
from typing import Any

SCHEMA = "mycellios-activation-sketch/1"
DEFAULT_SAMPLE_COUNT = 256
MAX_SAMPLE_COUNT = 1_024
MAX_ELEMENT_COUNT = 1 << 40
DEFAULT_COSINE_THRESHOLD = 0.99
DEFAULT_RELATIVE_NORM_THRESHOLD = 0.05
_SEED_BYTES = 16
_INDEX_DOMAIN = b"mycellios-activation-index/1\x00"
_COMMITMENT_DOMAIN = b"mycellios-activation-seed/1\x00"


class ActivationSketchError(ValueError):
    """The challenge or sketch is malformed and must fail closed."""


def generate_seed() -> str:
    """Return a fresh 128-bit verifier seed encoded as lowercase hex."""
    return secrets.token_hex(_SEED_BYTES)


def seed_commitment(seed: str) -> str:
    """Bind a seed before revealing which activation coordinates are sampled."""
    return "sha256:" + hashlib.sha256(_COMMITMENT_DOMAIN + _parse_seed(seed)).hexdigest()


def verify_seed_commitment(seed: str, commitment: str) -> bool:
    """Compare a revealed seed with its earlier commitment in constant time."""
    try:
        expected = seed_commitment(seed)
    except ActivationSketchError:
        return False
    return isinstance(commitment, str) and hmac.compare_digest(expected, commitment)


def projection_indices(seed: str, element_count: int,
                       sample_count: int = DEFAULT_SAMPLE_COUNT) -> tuple[int, ...]:
    """Derive device-independent, unique indexes from a bounded hash stream."""
    seed_bytes = _parse_seed(seed)
    if (not isinstance(element_count, int) or isinstance(element_count, bool)
            or element_count < 1 or element_count > MAX_ELEMENT_COUNT):
        raise ActivationSketchError("activation_element_count_is_invalid")
    if (not isinstance(sample_count, int) or isinstance(sample_count, bool)
            or sample_count < 1 or sample_count > min(MAX_SAMPLE_COUNT, element_count)):
        raise ActivationSketchError("activation_sample_count_is_invalid")
    selected: list[int] = []
    seen: set[int] = set()
    counter = 0
    while len(selected) < sample_count:
        digest = hashlib.sha256(
            _INDEX_DOMAIN + seed_bytes + counter.to_bytes(8, "big")
        ).digest()
        counter += 1
        index = int.from_bytes(digest[:8], "big") % element_count
        if index not in seen:
            seen.add(index)
            selected.append(index)
    return tuple(selected)


def make_activation_sketch(activation: Any, seed: str,
                           sample_count: int = DEFAULT_SAMPLE_COUNT) -> dict[str, Any]:
    """Create a bounded sketch from a Torch tensor without moving it wholesale."""
    try:
        import torch
    except ImportError as error:  # pragma: no cover - environment capability
        raise ActivationSketchError("activation_sketch_torch_is_unavailable") from error
    if not isinstance(activation, torch.Tensor):
        raise ActivationSketchError("activation_tensor_is_invalid")
    flattened = activation.detach().to(dtype=torch.float32).flatten()
    element_count = int(flattened.numel())
    indices = projection_indices(seed, element_count, sample_count)
    index_tensor = torch.tensor(indices, dtype=torch.long, device=flattened.device)
    projection = flattened.index_select(0, index_tensor).cpu().tolist()
    norm = float(torch.linalg.vector_norm(flattened).item())
    if not math.isfinite(norm) or any(not math.isfinite(float(value)) for value in projection):
        raise ActivationSketchError("activation_tensor_is_not_finite")
    return {
        "schema": SCHEMA,
        "seed": seed,
        "seedCommitment": seed_commitment(seed),
        "elementCount": element_count,
        "sampleCount": sample_count,
        "norm": norm,
        "projection": projection,
    }


def compare_activation_sketches(
    suspect: Mapping[str, Any], trusted: Mapping[str, Any], *,
    cosine_threshold: float = DEFAULT_COSINE_THRESHOLD,
    relative_norm_threshold: float = DEFAULT_RELATIVE_NORM_THRESHOLD,
) -> dict[str, Any]:
    """Compare two wire sketches without Torch; malformed input fails closed."""
    try:
        left = _parse_sketch(suspect)
        right = _parse_sketch(trusted)
        _validate_threshold(cosine_threshold, "activation_cosine_threshold_is_invalid")
        _validate_threshold(relative_norm_threshold, "activation_norm_threshold_is_invalid")
        if left["seed"] != right["seed"]:
            raise ActivationSketchError("activation_sketch_seed_mismatch")
        if left["seedCommitment"] != right["seedCommitment"]:
            raise ActivationSketchError("activation_sketch_commitment_mismatch")
        if (left["elementCount"] != right["elementCount"]
                or left["sampleCount"] != right["sampleCount"]):
            raise ActivationSketchError("activation_sketch_shape_mismatch")
        a, b = left["projection"], right["projection"]
        dot = math.fsum(x * y for x, y in zip(a, b, strict=True))
        magnitude_a = math.sqrt(math.fsum(x * x for x in a))
        magnitude_b = math.sqrt(math.fsum(y * y for y in b))
        cosine = (1.0 if magnitude_a == magnitude_b else 0.0) if (
            magnitude_a == 0 or magnitude_b == 0
        ) else max(-1.0, min(1.0, dot / (magnitude_a * magnitude_b)))
        denominator = max(left["norm"], right["norm"], sys.float_info.min)
        relative_norm = abs(left["norm"] - right["norm"]) / denominator
        passed = cosine >= cosine_threshold and relative_norm < relative_norm_threshold
        return {"passed": passed, "cosine": cosine, "relativeNorm": relative_norm,
                **({} if passed else {"error": "activation_sketch_diverged"})}
    except (ActivationSketchError, KeyError, TypeError, ValueError, OverflowError) as error:
        return {"passed": False, "cosine": 0.0, "relativeNorm": 1.0,
                "error": str(error) or "activation_sketch_is_malformed"}


def _parse_seed(seed: str) -> bytes:
    if not isinstance(seed, str) or len(seed) != _SEED_BYTES * 2:
        raise ActivationSketchError("activation_seed_is_invalid")
    try:
        decoded = bytes.fromhex(seed)
    except ValueError as error:
        raise ActivationSketchError("activation_seed_is_invalid") from error
    if len(decoded) != _SEED_BYTES or seed != seed.lower():
        raise ActivationSketchError("activation_seed_is_invalid")
    return decoded


def _parse_sketch(value: Mapping[str, Any]) -> dict[str, Any]:
    if not isinstance(value, Mapping) or value.get("schema") != SCHEMA:
        raise ActivationSketchError("activation_sketch_schema_is_invalid")
    seed, commitment = value.get("seed"), value.get("seedCommitment")
    if not verify_seed_commitment(seed, commitment):
        raise ActivationSketchError("activation_sketch_commitment_is_invalid")
    element_count, sample_count = value.get("elementCount"), value.get("sampleCount")
    projection, norm = value.get("projection"), value.get("norm")
    projection_indices(seed, element_count, sample_count)
    if (not isinstance(projection, Sequence)
            or isinstance(projection, (str, bytes, bytearray))
            or len(projection) != sample_count):
        raise ActivationSketchError("activation_sketch_projection_is_invalid")
    numbers = [float(item) for item in projection]
    norm_number = float(norm)
    if norm_number < 0 or not math.isfinite(norm_number):
        raise ActivationSketchError("activation_sketch_norm_is_invalid")
    if any(not math.isfinite(item) for item in numbers):
        raise ActivationSketchError("activation_sketch_projection_is_invalid")
    return {"seed": seed, "seedCommitment": commitment, "elementCount": element_count,
            "sampleCount": sample_count, "norm": norm_number, "projection": numbers}


def _validate_threshold(value: float, error: str) -> None:
    if (not isinstance(value, (int, float)) or isinstance(value, bool)
            or not math.isfinite(float(value)) or not 0 < float(value) <= 1):
        raise ActivationSketchError(error)


def main() -> int:
    """Read ``{suspect, trusted}`` JSON from stdin and emit a verdict."""
    try:
        request = json.load(sys.stdin)
        verdict = compare_activation_sketches(
            request["suspect"], request["trusted"],
            cosine_threshold=float(request.get("cosineThreshold", DEFAULT_COSINE_THRESHOLD)),
            relative_norm_threshold=float(request.get(
                "relativeNormThreshold", DEFAULT_RELATIVE_NORM_THRESHOLD)),
        )
    except (KeyError, TypeError, ValueError, json.JSONDecodeError) as error:
        json.dump({"passed": False,
                   "error": f"activation_challenge_request_invalid:{error}"}, sys.stdout)
        return 2
    json.dump(verdict, sys.stdout, separators=(",", ":"), sort_keys=True)
    return 0 if verdict["passed"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/contracts/activation-integrity.ts
import { createHash, timingSafeEqual } from "node:crypto";
import { z } from "zod";

export const ACTIVATION_SKETCH_SCHEMA = "mycellios-activation-sketch/1" as const;
export const ACTIVATION_INTEGRITY_EVIDENCE_SCHEMA =
  "mycellios-activation-integrity-evidence/1" as const;
export const DEFAULT_ACTIVATION_SAMPLE_COUNT = 256;
export const MAX_ACTIVATION_SAMPLE_COUNT = 1_024;
export const MAX_ACTIVATION_ELEMENT_COUNT = 2 ** 40;
export const DEFAULT_ACTIVATION_COSINE_THRESHOLD = 0.99;
export const DEFAULT_ACTIVATION_RELATIVE_NORM_THRESHOLD = 0.05;

const SEED_PATTERN = /^[0-9a-f]{32}$/;
const DIGEST_PATTERN = /^sha256:[0-9a-f]{64}$/;
const INDEX_DOMAIN = Buffer.from("mycellios-activation-index/1\0", "utf8");
const COMMITMENT_DOMAIN = Buffer.from("mycellios-activation-seed/1\0", "utf8");
const identifierSchema = z.string().regex(/^[A-Za-z0-9][A-Za-z0-9._:-]{0,255}$/);

export const activationSketchSchema = z.object({
  schema: z.literal(ACTIVATION_SKETCH_SCHEMA),
  seed: z.string().regex(SEED_PATTERN),
  seedCommitment: z.string().regex(DIGEST_PATTERN),
  elementCount: z.number().int().min(1).max(MAX_ACTIVATION_ELEMENT_COUNT),
  sampleCount: z.number().int().min(1).max(MAX_ACTIVATION_SAMPLE_COUNT),
  norm: z.number().nonnegative().finite(),
  projection: z.array(z.number().finite()).min(1).max(MAX_ACTIVATION_SAMPLE_COUNT),
}).strict().superRefine((sketch, context) => {
  if (sketch.sampleCount > sketch.elementCount) {
    context.addIssue({ code: "custom", path: ["sampleCount"], message: "activation_sample_count_is_invalid" });
  }
  if (sketch.projection.length !== sketch.sampleCount) {
    context.addIssue({ code: "custom", path: ["projection"], message: "activation_sketch_projection_is_invalid" });
  }
  if (!verifyActivationSeedCommitment(sketch.seed, sketch.seedCommitment)) {
    context.addIssue({ code: "custom", path: ["seedCommitment"], message: "activation_sketch_commitment_is_invalid" });
  }
});

export type ActivationSketch = z.infer<typeof activationSketchSchema>;

export interface ActivationSketchVerdict {
  passed: boolean;
  cosine: number;
  relativeNorm: number;
  error?: string;
}

export const activationIntegrityEvidenceSchema = z.object({
  schema: z.literal(ACTIVATION_INTEGRITY_EVIDENCE_SCHEMA),
  challengeId: identifierSchema,
  suspectStageId: identifierSchema,
  trustedStageId: identifierSchema,
  suspect: activationSketchSchema,
  trusted: activationSketchSchema,
  verdict: z.object({
    passed: z.literal(true),
    cosine: z.number().min(-1).max(1).finite(),
    relativeNorm: z.number().min(0).max(1).finite(),
  }).strict(),
}).strict().superRefine((evidence, context) => {
  if (evidence.suspectStageId === evidence.trustedStageId) {
    context.addIssue({ code: "custom", path: ["trustedStageId"], message: "activation_integrity_trusted_stage_must_be_distinct" });
  }
  const actual = compareActivationSketches(evidence.suspect, evidence.trusted);
  if (!actual.passed) {
    context.addIssue({ code: "custom", path: ["verdict"], message: actual.error ?? "activation_sketch_diverged" });
    return;
  }
  if (
    Math.abs(actual.cosine - evidence.verdict.cosine) > 1e-12
    || Math.abs(actual.relativeNorm - evidence.verdict.relativeNorm) > 1e-12
  ) {
    context.addIssue({ code: "custom", path: ["verdict"], message: "activation_integrity_verdict_mismatch" });
  }
});

export type ActivationIntegrityEvidence = z.infer<typeof activationIntegrityEvidenceSchema>;

export function activationSeedCommitment(seed: string): `sha256:${string}` {
  const seedBytes = parseSeed(seed);
  return `sha256:${createHash("sha256").update(COMMITMENT_DOMAIN).update(seedBytes).digest("hex")}`;
}

export function verifyActivationSeedCommitment(seed: string, commitment: string): boolean {
  if (!DIGEST_PATTERN.test(commitment)) return false;
  try {
    const expected = Buffer.from(activationSeedCommitment(seed), "utf8");
    const actual = Buffer.from(commitment, "utf8");
    return expected.length === actual.length && timingSafeEqual(expected, actual);
  } catch {
    return false;
  }
}

export function activationProjectionIndices(
  seed: string,
  elementCount: number,
  sampleCount = DEFAULT_ACTIVATION_SAMPLE_COUNT,
): number[] {
  const seedBytes = parseSeed(seed);
  if (!Number.isSafeInteger(elementCount) || elementCount < 1 || elementCount > MAX_ACTIVATION_ELEMENT_COUNT) {
    throw new Error("activation_element_count_is_invalid");
  }
  if (!Number.isSafeInteger(sampleCount) || sampleCount < 1 || sampleCount > Math.min(MAX_ACTIVATION_SAMPLE_COUNT, elementCount)) {
    throw new Error("activation_sample_count_is_invalid");
  }
  const indexes: number[] = [];
  const seen = new Set<number>();
  for (let counter = 0n; indexes.length < sampleCount; counter += 1n) {
    const counterBytes = Buffer.alloc(8);
    counterBytes.writeBigUInt64BE(counter);
    const digest = createHash("sha256")
      .update(INDEX_DOMAIN)
      .update(seedBytes)
      .update(counterBytes)
      .digest();
    const index = Number(digest.readBigUInt64BE(0) % BigInt(elementCount));
    if (!seen.has(index)) {
      seen.add(index);
      indexes.push(index);
    }
  }
  return indexes;
}

export function compareActivationSketches(
  suspectValue: unknown,
  trustedValue: unknown,
  options: { cosineThreshold?: number; relativeNormThreshold?: number } = {},
): ActivationSketchVerdict {
  try {
    const suspect = activationSketchSchema.parse(suspectValue);
    const trusted = activationSketchSchema.parse(trustedValue);
    const cosineThreshold = validThreshold(
      options.cosineThreshold ?? DEFAULT_ACTIVATION_COSINE_THRESHOLD,
      "activation_cosine_threshold_is_invalid",
    );
    const normThreshold = validThreshold(
      options.relativeNormThreshold ?? DEFAULT_ACTIVATION_RELATIVE_NORM_THRESHOLD,
      "activation_norm_threshold_is_invalid",
    );
    if (suspect.seed !== trusted.seed) throw new Error("activation_sketch_seed_mismatch");
    if (suspect.seedCommitment !== trusted.seedCommitment) throw new Error("activation_sketch_commitment_mismatch");
    if (suspect.elementCount !== trusted.elementCount || suspect.sampleCount !== trusted.sampleCount) {
      throw new Error("activation_sketch_shape_mismatch");
    }
    let dot = 0;
    let magnitudeSuspect = 0;
    let magnitudeTrusted = 0;
    for (let index = 0; index < suspect.sampleCount; index += 1) {
      const left = suspect.projection[index]!;
      const right = trusted.projection[index]!;
      dot += left * right;
      magnitudeSuspect += left * left;
      magnitudeTrusted += right * right;
    }
    magnitudeSuspect = Math.sqrt(magnitudeSuspect);
    magnitudeTrusted = Math.sqrt(magnitudeTrusted);
    const cosine = magnitudeSuspect === 0 || magnitudeTrusted === 0
      ? (magnitudeSuspect === magnitudeTrusted ? 1 : 0)
      : Math.max(-1, Math.min(1, dot / (magnitudeSuspect * magnitudeTrusted)));
    const relativeNorm = Math.abs(suspect.norm - trusted.norm)
      / Math.max(suspect.norm, trusted.norm, Number.MIN_VALUE);
    const passed = cosine >= cosineThreshold && relativeNorm < normThreshold;
    return { passed, cosine, relativeNorm, ...(passed ? {} : { error: "activation_sketch_diverged" }) };
  } catch (error) {
    return {
      passed: false,
      cosine: 0,
      relativeNorm: 1,
      error: error instanceof Error && error.message ? error.message : "activation_sketch_is_malformed",
    };
  }
}

function parseSeed(seed: string): Buffer {
  if (!SEED_PATTERN.test(seed)) throw new Error("activation_seed_is_invalid");
  return Buffer.from(seed, "hex");
}

function validThreshold(value: number, error: string): number {
  if (!Number.isFinite(value) || value <= 0 || value > 1) throw new Error(error);
  return value;
}
